stdlib/private/OSLog/OSLogMessage.swift

```diff
@@ -0,0 +1,285 @@
+  public init(stringLiteral value: String) {
+    // Note that the actual value of `literalCapacity` is not important as it
+    // is ignored by `OSLogInterpolation.init`. However, it must be a literal.
+    var s = OSLogInterpolation(literalCapacity: 1, interpolationCount: 0)
+    s.appendLiteral(value)
+    self.init(stringInterpolation: s)
+  }
+
+  /// Format string constructed from the string interpolation.
+  public var formatString: String {
+    get { return oslogInterpolation.formatString }
+  }
+
+  /// The byte size of the argument buffer that will contain the preamble
+  /// and the elements of `oslogInterpolation.argumentSequence`.
+  public var argumentBufferByteSize: Int {
+    get { return oslogInterpolation.argumentSequence.byteCount + 2 }
+  }
+
+  /// Serialize the summary bytes and arguments into the given byte-buffer
+  /// builder.
+  public func serializeArguments(
+    into bufferBuilder: inout OSLogByteBufferBuilder) {
+    bufferBuilder.serialize(oslogInterpolation.preamble)
+    bufferBuilder.serialize(oslogInterpolation.argumentCount)
+    oslogInterpolation.argumentSequence.serialize(into: &bufferBuilder)
+  }
+}
+
+/// A representation of a sequence of arguments and headers (of possibly
+/// different types) that have to be serialized to a byte buffer. The arguments
+/// are captured within closures and stored in an array. The closures accept an
+/// instance of `OSLogByteBufferBuilder`, and when invoked, serialize the
+/// argument using the passed `OSLogByteBufferBuilder` instance.
+public struct OSLogArgumentSequence {
+  /// An array of closures that captures arguments of possibly different types.
+  public var argumentClosures: [(inout OSLogByteBufferBuilder) -> ()]
+  /// Sum total of the byte size of the arguments that are tracked.
+  public var byteCount: Int
+
+  public init() {
+    argumentClosures = []
+    byteCount = 0
+  }
+
+  /// Append a byte-sized header, constructed by
+  /// `OSLogMessage.appendInterpolation`, to the tracked array of closures.
+  public mutating func append(_ header: UInt8) {
+    argumentClosures.append({ $0.serialize(header) })
+    byteCount += OSLogByteBufferBuilder.numBytesForEncoding(UInt8.self)
+  }
+
+  /// Append an (autoclosured) interpolated expression of type Int, passed to
+  /// `OSLogMessage.appendInterpolation`, to the tracked array of closures.
+  public mutating func append(_ value: @escaping () -> Int) {
+    argumentClosures.append({ $0.serialize(value()) })
+    byteCount += OSLogByteBufferBuilder.numBytesForEncoding(Int.self)
+  }
+
+  public func serialize(into bufferBuilder: inout OSLogByteBufferBuilder) {
+    argumentClosures.forEach { $0(&bufferBuilder) }
+  }
+}
+
+/// A struct that manages serialization of instances of specific types to
+/// a byte buffer. The byte buffer is provided as an argument to the initializer
+/// so that its lifetime can be managed by the caller.
+public struct OSLogByteBufferBuilder {
```

Comment:
Looks like an implementation detail, can it be made `internal` or even `private`?

Reply:
It kind of has the same problem as the `OSLogArguments`. They have to be extendable for adding support for serializing custom types. Every overload of `serialize` method on this type defines how to marshal a type into bytes.

Reply:
Is there a set of primitive types that custom types' serialization can be described in terms of?

Reply:
@jckarter  That is a good point. Yeah, there is.  The primitives types supported by`os_log` are Ints, Floats, Strings, Pointers and NSObject. If we add NSObject serialization also in this same OSLog module, it will create a dependency on Foundation. But, I guess that is okay.  So, in principle, both `OSLogArguments` and `OSLogByteBufferBuilder` could be made internal.

Reply:
I have made a commit in which `OSLogByteBufferBuilder` is internal, and `OSLogArguments` is public but its properties are internal. Since `OSLogArguments.append` is used by string interpolation functions, which can be customized from outside the module, I have kept that type and those methods public.